internal/pkg/docker/orchestrator/orchestrator.go

```diff
@@ -452,6 +486,86 @@ func (o *Orchestrator) waitForContainerToStart(ctx context.Context, id string) e
 	}
 }
 
+func (o *Orchestrator) waitForContainerDependencies(ctx context.Context, name string, definitions map[string]ContainerDefinition) error {
+	var deps []string
+	for depName, state := range definitions[name].DependsOn {
+		deps = append(deps, fmt.Sprintf("%s->%s", depName, state))
+	}
+	logMsg := strings.Join(deps, ", ")
+	fmt.Printf("Waiting for container %q dependencies: [%s]\n", name, color.Emphasize(logMsg))
+	eg, ctx := errgroup.WithContext(ctx)
+	for name, state := range definitions[name].DependsOn {
+		name, state := name, state
+		eg.Go(func() error {
+			ctrId := o.containerID(name)
+			isEssential := definitions[name].IsEssential
+			ticker := time.NewTicker(700 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-ticker.C:
+				case <-ctx.Done():
+					return ctx.Err()
+				}
+				switch state {
+				case ctrStateStart:
+					err := o.waitForContainerToStart(ctx, ctrId, isEssential)
+					if err != nil {
+						return err
+					}
+					if isEssential {
+						log.Successf("Successfully started container %s\n", ctrId)
+						return nil
+					}
+					fmt.Printf("non-essential container %q started successfully\n", ctrId)
+					return nil
+				case ctrStateHealthy:
+					healthy, err := o.docker.IsContainerHealthy(ctx, ctrId)
+					if err != nil {
+						if !isEssential {
+							fmt.Printf("non-essential container %q is not healthy: %v\n", ctrId, err)
+							return nil
+						}
+						return fmt.Errorf("essential container %q is not healthy: %w", ctrId, err)
+					}
+					if healthy {
+						log.Successf("Successfully dependency container %q reached %q\n", ctrId, state)
+						return nil
+					}
+				case ctrStateComplete:
+					exitCode, err := o.docker.ContainerExitCode(ctx, ctrId)
+					if errors.Is(err, fmt.Errorf("container %q has not exited", name)) {
```

Comment:
let's make `fmt.Errorf("container %q has not exited", name)` a typed error so that we can check by `errors.As` here! We could end up changing the error message in `dockerenginer` and that would render this error check useless 😢